pkg/distributor/distributor.go

```diff
@@ -150,6 +151,11 @@ type Distributor struct {
 
 	// Pool of []byte used when marshalling write requests.
 	writeRequestBytePool sync.Pool
+
+	// ingesterDoBatchPushGo is the Go function passed to ring.DoBatchWithOptions.
+	// It can be nil, in which case a simple `go f()` will be used.
+	// See Config.IngesterPushWorkerGoroutines on how to configure this.
+	ingesterDoBatchPushGo func(func())
```

Comment:
nit: Can we use something else besides `Go` here? I'm struggling to understand what this is/does.

Reply:
Renamed to `ingesterDoBatchPushWorkers`, does that work? 
I called it `Go` because that's how I called the option in `dskit`, I'm also open to suggestion on naming there.

Reply:
Since the new CLI flag is called `distributor.reusable-ingester-push-workers`, my suggestion is:
- replace 
   ```
     ingesterDoBatchPushWorkers func(func())
   ``` 
   with 
   ```
     reusableIngesterPushWorkersPool *concurrency.ReusableGoroutinesPool
   ```
- instantiate it as
   ```
     if cfg.ReusableIngesterPushWorkers > 0 {
       d.reusableIngesterPushWorkersPool := concurrency.NewReusableGoroutinesPool(cfg.ReusableIngesterPushWorkers)
       ...
	}
   ```
- and use it as
   ```
     ring.DoBatchOptions{
       Cleanup:       func() { pushReq.CleanUp(); cancel() },
       IsClientError: isClientError,
       Go:            d.reusableIngesterPushWorkersPool.Go,
     }
   ```

It is more intuitive what the type `*concurrency.ReusableGoroutinesPool` is comparing to just `func(func())`.

Reply:
But then I need to add extra logic on each call checking whether there's a pool or not.

Reply:
I don't want to block you on this. I approved your PR.
If we don't manage to find a better attribute name, let's merge it like this.